### DISCUSSION TITLE
Type checking failure / weird resolution

### DISCUSSION THREAD

**Discussion by oskarirauta** [nlohmann/json#3982] [General]
data contains following: `{"color":"#96C2E5","font":"regular","size":23,"spacing":3}`
and type is `nlohmann::json`

Here's the code:
```
        if ( data.contains("font") && !data["font"].is_string())
                return utils::JSON_INVALID_FONT;
        if ( data.contains("size") && !data["size"].is_number_unsigned())
                return utils::JSON_INVALID_SIZE;
        if ( data.contains("spacing") && !data["spacing"].is_number_unsigned())
                return utils::JSON_INVALID_SPACING;
        if ( !data.contains("rotate") && !data["rotate"].is_number_unsigned())
                return utils::JSON_INVALID_ROTATE;
        if ( data.contains("enlarge") && !data["enlarge"].is_boolean())
                return utils::JSON_INVALID_ENLARGE;
        if ( data.contains("stretch") && !data["stretch"].is_boolean())
                return utils::JSON_INVALID_STRETCH;
        if ( data.contains("x2") && !data["x2"].is_number_integer())
                return utils::JSON_INVALID_X2;
        if ( data.contains("y2") && !data["y2"].is_number_integer())
                return utils::JSON_INVALID_Y2;
        if ( data.contains("thickness") && !data["thickness"].is_number_unsigned())
                return utils::JSON_INVALID_THICKNESS;
        if ( data.contains("radius") && !data["radius"].is_number_unsigned())
                return utils::JSON_INVALID_RADIUS;
        if ( data.contains("width") && !data["width"].is_number_integer())
                return utils::JSON_INVALID_WIDTH;
        if ( data.contains("height") && !data["height"].is_number_integer())
                return utils::JSON_INVALID_HEIGHT;
        if ( data.contains("color") && !data["color"].is_string())
                return utils::JSON_INVALID_COLOR;
        if ( data.contains("fill") && !data["fill"].is_string())
                return utils::JSON_INVALID_FILL;
        if ( data.contains("background") && !data["background"].is_string())
                return utils::JSON_INVALID_BACKGROUND;
```

and here's result:
```
user@media:/usr/src/splash2# ./splash2 --msg hello
Work in progress..
terminate called after throwing an instance of 'nlohmann::json_abi_v3_11_2::detail::type_error'
  what():  [json.exception.type_error.302] type must be string, but is null
zsh: abort      ./splash2 --msg hello
user@media:/usr/src/splash2#
```

My validation routine is long, so partially including it only. For part that causes the problem.
If I comment out that section, program works as expected (except that it no longer checks that valid format was used on behalf on value types).

My program supports few functions; like msg, img and shape. If I use different functions, different kind of properties are required.. And tested.
With image type I got result that must be string, but is null and commenting out this made it go through:
```
        if ( data.contains("enlarge") && !data["enlarge"].is_boolean())
                return utils::JSON_INVALID_ENLARGE;
```

Why would value need to be anything there? Or why exception, when string doesn't have anything to do with that. Or with msg, so what if something is null - we return our own error then to caller..

Any ideas?

**Comment by nlohmann** [comment]
Which line throws the exception?

**Reply by oskarirauta** [reply]
It is different if json is different, program is supposed to be those "boot splash" programs displaying content on screen during boot, so you can have one of following: image, text, shape

settings are in json, and type and provided json alter "data" json of course..
With empty settings and type image, problem was this line:

```
if ( !data.contains("rotate") && !data["rotate"].is_number_unsigned())
                return utils::JSON_INVALID_ROTATE;
```

But for message(text), data contains different entries, and then it's other row (I didn't match with one it is..)

edit: It is actually same line again. Always seems to be line that checks rotation options.

**Reply by gregmarr** [reply]
The line has an extra `!` at the beginning.  You're checking the type if it's NOT there.

**Reply by oskarirauta** [reply]
Wow! Thanks :D I should go to eye doctor :D

**Reply by gregmarr** [reply]
You're welcome.

**Comment by oskarirauta** [comment]
Resolved by @gregmarr

user error due to typo :D
